Windows as target platform

As games who want to be released through [Steam Greenlight](https://steamcommunity.com/greenlight/faq/), Windows support is kinda a must. 

However, `engi` currently doesn't build with `GOOS=windows`.

Build errors:
```
../../../webgl/webgl_gl2.go:15:2: no buildable Go source files in /home/etiennebruines/workspaces/go/src/github.com/go-gl/gl/v2.1/gl
../../engi_glfw.go:20:2: no buildable Go source files in /home/etiennebruines/workspaces/go/src/github.com/go-gl/glfw/v3.1/glfw
../../audio.go:7:2: no buildable Go source files in /home/etiennebruines/workspaces/go/src/golang.org/x/mobile/exp/audio/al
```

--- a/audio_windows.go
+++ b/audio_windows.go
@@ -1,0 +1,41 @@
+package engi
+
+import (
+	"log"
+
+	"github.com/paked/engi/ecs"
+)
+
+const (
+	defaultHeightModifier float32 = 1
+)
+
+// AudioComponent is a Component which is used by the AudioSystem
+type AudioComponent struct {
+	File       string
+	Repeat     bool
+	Background bool
+	player     *Player
+}
+
+func (*AudioComponent) Type() string {
+	return "AudioComponent"
+}
+
+// AudioSystem is a System that allows for sound effects and / or music
+type AudioSystem struct {
+	*ecs.System
+	HeightModifier float32
+}
+
+func (AudioSystem) Type() string {
+	return "AudioSystem"
+}
+
+func (as *AudioSystem) New(*ecs.World) {
+	as.System = ecs.NewSystem()
+
+	log.Println("Warning: audio is not yet implemented on Windows")
+}
+
+func (as *AudioSystem) Update(entity *ecs.Entity, dt float32) {}

--- a/batch.go
+++ b/batch.go
@@ -78,8 +78,7 @@ func NewBatch(width, height float32, vertSrc, fragSrc string) *Batch {
 	Gl.VertexAttribPointer(batch.inTexCoords, 2, Gl.FLOAT, false, 20, 8)
 	Gl.VertexAttribPointer(batch.inColor, 4, Gl.UNSIGNED_BYTE, true, 20, 16)
 
-	batch.projX = width
-	batch.projY = height
+	batch.SetProjection(width, height)
 
 	Gl.Enable(Gl.BLEND)
 	Gl.BlendFunc(Gl.SRC_ALPHA, Gl.ONE_MINUS_SRC_ALPHA)
@@ -115,7 +114,7 @@ func (b *Batch) flush() {
 	Gl.BindTexture(Gl.TEXTURE_2D, b.lastTexture)
 
 	Gl.Uniform2f(b.ufProjection, b.projX, b.projY)
-	Gl.Uniform3f(b.center, cam.x/Width(), cam.y/Height(), cam.z)
+	Gl.Uniform3f(b.center, cam.x/b.projX, cam.y/b.projY, cam.z)
 
 	Gl.BufferSubData(Gl.ARRAY_BUFFER, 0, b.vertices[0:b.index*20])
 	Gl.DrawElements(Gl.TRIANGLES, 6*b.index, Gl.UNSIGNED_SHORT, 0)
@@ -123,9 +122,9 @@ func (b *Batch) flush() {
 	b.index = 0
 }
 
-func (b *Batch) SetProjection(width, height, depth float32) {
-	b.projX = width
-	b.projY = height
+func (b *Batch) SetProjection(width, height float32) {
+	b.projX = width / 2
+	b.projY = height / 2
 }
 
 func (b *Batch) Draw(r Drawable, x, y, originX, originY, scaleX, scaleY, rotation float32, color color.Color, transparency float32) {

--- a/demos/keyboardscroller/kbscroller.go
+++ b/demos/keyboardscroller/kbscroller.go
@@ -68,5 +68,5 @@ func (game *Game) Setup(w *ecs.World) {
 }
 
 func main() {
-	engi.Open("KeyboardScroller Demo", 400, 400, false, &Game{})
+	engi.Open("KeyboardScroller Demo", int(worldWidth), int(worldHeight), false, &Game{})
 }

--- a/demos/mouse/README.md
+++ b/demos/mouse/README.md
@@ -1,0 +1,10 @@
+# Mouse Demo
+
+## What does it do?
+It demonstrates how one can use mouse events in `engi`. 
+
+## What are important aspects of the code?
+These lines are key in this demo:
+
+* `entity.AddComponent(&engi.MouseComponent{})` which adds a `MouseComponent` to the `Entity` of which we want to know mouse stuff; 
+* `w.AddSystem(&engi.MouseSystem{})` which adds the `MouseSystem` to this `Scene`;

--- a/demos/mouse/mouse.go
+++ b/demos/mouse/mouse.go
@@ -1,0 +1,89 @@
+package main
+
+import (
+	"github.com/paked/engi"
+	"github.com/paked/engi/ecs"
+	"image"
+	"image/color"
+)
+
+var (
+	World     *GameWorld
+	boxWidth  float32 = 150
+	boxHeight float32 = 150
+)
+
+type GameWorld struct{}
+
+func (game *GameWorld) Preload() {}
+
+// generateBackground creates a background of green tiles - might not be the most efficient way to do this
+func generateBackground() *engi.RenderComponent {
+	rect := image.Rect(0, 0, int(boxWidth), int(boxHeight))
+	img := image.NewNRGBA(rect)
+	c1 := color.RGBA{102, 153, 0, 255}
+	for i := rect.Min.X; i < rect.Max.X; i++ {
+		for j := rect.Min.Y; j < rect.Max.Y; j++ {
+			img.Set(i, j, c1)
+		}
+	}
+	bgTexture := engi.NewImageObject(img)
+	fieldRender := engi.NewRenderComponent(engi.NewRegion(engi.NewTexture(bgTexture), 0, 0, int(boxWidth), int(boxHeight)), engi.Point{1, 1}, "Background1")
+	fieldRender.SetPriority(engi.Background)
+	return fieldRender
+}
+
+func (game *GameWorld) Setup(w *ecs.World) {
+	engi.SetBg(0xFFFFFF)
+
+	w.AddSystem(&engi.MouseSystem{})
+	w.AddSystem(&engi.RenderSystem{})
+	w.AddSystem(&ControlSystem{})
+	w.AddSystem(engi.NewMouseZoomer(-0.125))
+
+	w.AddEntity(game.CreateEntity())
+}
+
+func (*GameWorld) Hide()        {}
+func (*GameWorld) Show()        {}
+func (*GameWorld) Type() string { return "GameWorld" }
+
+func (game *GameWorld) CreateEntity() *ecs.Entity {
+	entity := ecs.NewEntity([]string{"MouseSystem", "RenderSystem", "ControlSystem"})
+
+	entity.AddComponent(generateBackground())
+	entity.AddComponent(&engi.MouseComponent{})
+	entity.AddComponent(&engi.SpaceComponent{engi.Point{0, 0}, boxWidth, boxHeight})
+
+	return entity
+}
+
+type ControlSystem struct {
+	*ecs.System
+}
+
+func (ControlSystem) Type() string {
+	return "ControlSystem"
+}
+
+func (c *ControlSystem) New(*ecs.World) {
+	c.System = ecs.NewSystem()
+}
+
+func (c *ControlSystem) Update(entity *ecs.Entity, dt float32) {
+	var mouse *engi.MouseComponent
+	if !entity.Component(&mouse) {
+		return
+	}
+
+	if mouse.Enter {
+		engi.SetCursor(engi.Hand)
+	} else if mouse.Leave {
+		engi.SetCursor(engi.Arrow)
+	}
+}
+
+func main() {
+	World = &GameWorld{}
+	engi.Open("Mouse Demo", 1500, 400, false, World)
+}

--- a/engi.go
+++ b/engi.go
@@ -21,8 +21,9 @@ var (
 	currentWorld *ecs.World
 	currentScene Scene
 
+	scaleOnResize   = false
 	fpsLimit        = 120
-	headless        bool
+	headless        = false
 	resetLoopTicker = make(chan bool, 1)
 )
 
@@ -58,6 +59,10 @@ func SetBg(color uint32) {
 		b := float32(color&0xFF) / 255.0
 		Gl.ClearColor(r, g, b, 1.0)
 	}
+}
+
+func SetScaleOnResize(b bool) {
+	scaleOnResize = b
 }
 
 func SetFPSLimit(limit int) error {

--- a/engi_glfw.go
+++ b/engi_glfw.go
@@ -33,8 +33,10 @@ var (
 	HResize   *glfw.Cursor
 	VResize   *glfw.Cursor
 
-	headlessWidth  = 800
-	headlessHeight = 800
+	headlessWidth             = 800
+	headlessHeight            = 800
+	gameWidth, gameHeight     float32
+	windowWidth, windowHeight float32
 )
 
 // fatalErr calls log.Fatal with the given error if it is non-nil.
@@ -56,6 +58,9 @@ func run(defaultScene Scene, title string, width, height int, fullscreen bool) {
 
 	monitor := glfw.GetPrimaryMonitor()
 	mode := monitor.GetVideoMode()
+
+	gameWidth = float32(width)
+	gameHeight = float32(height)
 
 	if fullscreen {
 		width = mode.Width
@@ -83,9 +88,11 @@ func run(defaultScene Scene, title string, width, height int, fullscreen bool) {
 
 	Gl = webgl.NewContext()
 	Gl.Viewport(0, 0, width, height)
+
 	window.SetFramebufferSizeCallback(func(window *glfw.Window, w, h int) {
 		width, height = window.GetFramebufferSize()
 		Gl.Viewport(0, 0, width, height)
+
 		// TODO: when do we want to handle resizing? and who should deal with it?
 		// responder.Resize(w, h)
 	})
@@ -117,6 +124,24 @@ func run(defaultScene Scene, title string, width, height int, fullscreen bool) {
 			keyStates[key] = true
 		} else if a == glfw.Release {
 			keyStates[key] = false
+		}
+	})
+
+	window.SetSizeCallback(func(w *glfw.Window, widthInt int, heightInt int) {
+		windowWidth = float32(widthInt)
+		windowHeight = float32(heightInt)
+
+		if !scaleOnResize {
+			gameWidth, gameHeight = float32(widthInt), float32(heightInt)
+
+			// Update default batch
+			for _, world := range worlds {
+				for _, s := range world.Systems() {
+					if render, ok := s.(*RenderSystem); ok {
+						render.defaultBatch.SetProjection(gameWidth, gameHeight)
+					}
+				}
+			}
 		}
 	})
 
@@ -187,23 +212,19 @@ Outer:
 }
 
 func Width() float32 {
-	if headless {
-		return float32(headlessWidth)
-	}
-	width, _ := window.GetSize()
-	return float32(width)
+	return gameWidth
 }
 
 func Height() float32 {
-	if headless {
-		return float32(headlessHeight)
-	}
-	_, height := window.GetSize()
-	return float32(height)
+	return gameHeight
 }
 
 func Exit() {
 	window.SetShouldClose(true)
+}
+
+func SetCursor(c *glfw.Cursor) {
+	window.SetCursor(c)
 }
 
 func init() {

--- a/mouse.go
+++ b/mouse.go
@@ -1,0 +1,121 @@
+package engi
+
+import (
+	"github.com/paked/engi/ecs"
+)
+
+// MouseComponent is the location for the MouseSystem to store its results; to be used / viewed by other Systems
+type MouseComponent struct {
+	// Clicked is true whenever the Mouse was space in this frame
+	Clicked bool
+	// Hovered is true whenever the Mouse is hovering the space in this frame
+	Hovered bool
+	// Dragged is true whenever the space was clicked, and then the mouse started moving (while holding)
+	Dragged bool
+	// Rightclicked is true whenever the space was rightclicked in this frame
+	Rightclicked bool
+	// Enter is true whenever the Mouse entered the space in that frame, but wasn't in that space on the previous frame
+	Enter bool
+	// Leave is true whenever the Mouse was in the space on the previous frame, but now isn't
+	Leave bool
+}
+
+// Type returns the string representation of the MouseComponent type
+func (*MouseComponent) Type() string {
+	return "MouseComponent"
+}
+
+// MouseSystem listens for mouse events, and changes value for MouseComponent accordingly
+type MouseSystem struct {
+	*ecs.System
+
+	mouseX    float32
+	mouseY    float32
+	mouseDown bool
+}
+
+// Type returns the string representation of the MouseSystem type
+func (*MouseSystem) Type() string {
+	return "MouseSystem"
+}
+
+// New initializes the MouseSystem
+func (m *MouseSystem) New(*ecs.World) {
+	m.System = ecs.NewSystem()
+}
+
+// Priority returns a priority of 10 (higher than most) to ensure that this System runs before all others
+func (m *MouseSystem) Priority() int {
+	return 10
+}
+
+// Pre is called before all Update calls, and is used to compute internal values
+func (m *MouseSystem) Pre() {
+	// Translate Mouse.X and Mouse.Y into "game coordinates"
+	m.mouseX = Mouse.X*cam.z*(gameWidth/windowWidth) + cam.x - (gameWidth/2)*cam.z
+	m.mouseY = Mouse.Y*cam.z*(gameHeight/windowHeight) + cam.y - (gameHeight/2)*cam.z
+}
+
+// Update sets the MouseComponent values for each Entity
+func (m *MouseSystem) Update(entity *ecs.Entity, dt float32) {
+	var (
+		mc     *MouseComponent
+		space  *SpaceComponent
+		render *RenderComponent
+		ok     bool
+	)
+
+	// We need MouseComponent to save our findings
+	if mc, ok = entity.ComponentFast(mc).(*MouseComponent); !ok {
+		return
+	}
+
+	// We need SpaceComponent for the location
+	if space, ok = entity.ComponentFast(space).(*SpaceComponent); !ok {
+		return
+	}
+
+	// We need RenderComponent for the Priority
+	if render, ok = entity.ComponentFast(render).(*RenderComponent); !ok {
+		return
+	}
+
+	// Reset some values
+	mc.Leave = false
+
+	mx := m.mouseX
+	my := m.mouseY
+
+	// Special case: HUD
+	if render.priority >= HUDGround {
+		mx = Mouse.X
+		my = Mouse.Y
+	}
+
+	// Check if the X-value is within range
+	// and if the Y-value is within range
+	if mx > space.Position.X && mx < (space.Position.X+space.Width) &&
+		my > space.Position.Y && my < (space.Position.Y+space.Height) {
+
+		mc.Enter = !mc.Hovered
+		mc.Hovered = true
+
+		switch Mouse.Action {
+		case PRESS:
+			mc.Clicked = true
+			m.mouseDown = true
+		case RELEASE:
+			m.mouseDown = false
+			mc.Dragged = false
+		case MOVE:
+			if m.mouseDown {
+				mc.Dragged = true
+			}
+		}
+	} else {
+		if mc.Hovered {
+			mc.Leave = true
+		}
+		mc.Hovered = false
+	}
+}

